src/runtime/rpc/rpc_address.cpp

```diff
@@ -68,23 +68,65 @@ error_s rpc_address::GetAddrInfo(const std::string &hostname, const addrinfo &hi
     return error_s::ok();
 }
 
+bool extract_host_port(const std::string_view &host_port, std::string_view &host, uint16_t &port)
+{
+    // Check the port field is present.
+    const auto pos = host_port.find_last_of(':');
+    if (dsn_unlikely(pos == std::string::npos)) {
+        LOG_ERROR("bad format, should be in the form of <ip|host>:port, but got '{}'", host_port);
+        return false;
+    }
+
+    // Check port.
+    const auto port_str = host_port.substr(pos + 1);
+    if (dsn_unlikely(!dsn::buf2uint16(port_str, port))) {
+        LOG_ERROR("bad port, should be an uint16_t integer, but got '{}'", port_str);
+        return false;
+    }
+
+    host = host_port.substr(0, pos);
+    return true;
+}
+
 const rpc_address rpc_address::s_invalid_address;
 
+rpc_address::rpc_address(uint32_t ip, uint16_t port)
+{
+    _addr.v4.type = HOST_TYPE_IPV4;
+    _addr.v4.ip = ip;
+    _addr.v4.port = port;
+    static_assert(sizeof(rpc_address) == sizeof(uint64_t),
+                  "make sure rpc_address does not add new payload to "
+                  "rpc_address to keep it sizeof(uint64_t)");
+}
+
+rpc_address::rpc_address(const struct sockaddr_in &addr)
+{
+    _addr.v4.type = HOST_TYPE_IPV4;
+    _addr.v4.ip = static_cast<uint32_t>(ntohl(addr.sin_addr.s_addr));
+    _addr.v4.port = ntohs(addr.sin_port);
+}
+
+rpc_address::rpc_address(const rpc_address &another) { *this = another; }
```

Comment:
Use `rpc_address(const rpc_address &) = default;` instead ?

Reply:
Done

Reply:
It doesn't works. Keep the code to explict using `=` operator which has been overwriten, it contains a `add_ref()` operation to ensure the ref counter. But if using default, it is copied bitwise, the ref counter is lost.